### DISCUSSION TITLE
Property 'verified' is not set after successful signature verification

### DISCUSSION THREAD

**Discussion by IvanPizhenko** [openpgpjs/openpgpjs#1491] [Q&A]
- OpenPGP.js version: 5.1.0
- Affected platform (Browser or Node.js version): Node 12 (v12.22.10)

Following code in `async verify(key, signatureType, data, ...`
```js
this[verified] = await mod.signature.verify( ...
```

breaks following use case (Typescript, source code origin: https://github.com/FlowCrypt/flowcrypt-ios/, please note: this code is "source available", but not open source):
```typescript
  public static lastSig = async (key: Key): Promise<number> => {
    let i = 0;
    console.log(`Key OUID ${(Object as any).__oid(key)}`);
    await key.getExpirationTime(); // will force all sigs to be verified
    const allSignatures: SignaturePacket[] = [];
    for (const user of key.users) {
      i = 0;
      for (const sig of user.selfCertifications) {
        console.log(`user ${user.userID?.email} sig[${i++}](OUID ${(Object as any).__oid(sig)} => ${sig.verified})\n`);
      }
      allSignatures.push(...user.selfCertifications);
    }
    i = 0;
    for (const subKey of key.subkeys) {
      i = 0;
      for (const sig of subKey.bindingSignatures) {
        console.log(`subkey ${subKey.getKeyID().toHex()} sig[${i++}](OUID ${(Object as any).__oid(sig)} => ${sig.verified})}\n`);
      }
      allSignatures.push(...subKey.bindingSignatures);
    }
    allSignatures.sort((a, b) => (b.created ? b.created.getTime() : 0) - (a.created ? a.created.getTime() : 0));
    console.log(`nsig=${allSignatures.length}`);
    i = 0;
    for (const sig of allSignatures) {
      console.log(`sig[${i++}](OUID ${(Object as any).__oid(sig)} => ${sig.verified})}\n`);
    }
    const newestSig = allSignatures.find(sig => sig.verified === true);
    if (newestSig) {
      return newestSig.created ? newestSig.created.getTime() : 0;
    }
    throw new Error('No valid signature found in key');
  }
```
After calling `await key.getExpirationTime(); // will force all sigs to be verified` we expect signatures to be verified, however when enumerating signatures, `verified` is not set.
I could solve it by changing all `this[verified]` into `this.verified`.

Additional info:
Extra stuff used here:
```javascript
// Based on https://stackoverflow.com/a/1997811/1540501
// inject oids
(function() {
  if ( typeof Object.__oid == "undefined" ) {
      var id = 0;

      Object.__oid = function(o) {
          if ( typeof o.__uniqueid == "undefined" ) {
              Object.defineProperty(o, "__uniqueid", {
                  value: ++id,
                  enumerable: false,
                  // This could go either way, depending on your
                  // interpretation of what an "id" is
                  writable: false
              });
          }

          return o.__uniqueid;
      };
  }
})();

```

**Comment by larabr** [comment]
The verified flag was converted to a symbol in v5, and it is not meant to be externally accessible.
The reason is to avoid misusing the flag. The value is now cached and taken into account (when possible) as part of `signature.verify`. Hence, if you call `verify` on an already verified signature, the operation should be fairly cheap. See for example what we do here: https://github.com/openpgpjs/openpgpjs/blob/main/src/key/helper.js#L54

That said, [we've also changed `Key.getExpirationTime`](https://github.com/openpgpjs/openpgpjs/pull/1319), and it no longer checks subkeys, but only returns primary key expiration. As a result, it is no longer true that it "will force all sigs to be verified".
In general, I think it's best if your functions do not depend on side effects of what OpenPGP.js does, because it can become very difficult to upgrade as a result, as the internal structures of the library can/will change.

What's the intended use for the result of `lastSig()`? We could consider exposing a function that does the same, if it's of general interest. Or I can help you find another way to implement the function.

**Reply by IvanPizhenko** [reply]
> no longer checks subkeys, but only returns primary key expiration

yes, that may be significant. The only use I see now is getting date/time of last key modification as date of that last signature. So can you please suggest any other way how to implement this?

**Reply by IvanPizhenko** [reply]
> What's the intended use for the result of lastSig()? We could consider exposing a function that does the same, if it's of general interest. Or I can help you find another way to implement the function.

Yes, that would be ideal if library provides such function OOB. Is it possible to do this in the nearest time?

**Reply by IvanPizhenko** [reply]
> In general, I think it's best if your functions do not depend on side effects of what OpenPGP.js does, because it can become very difficult to upgrade as a result, as the internal structures of the library can/will change.

Obviously it is better, but (a) this is some legacy code (b) typescript definition is misleading, it still say that some things are still public available (see my comment below).

**Reply by larabr** [reply]
> The only use I see now is getting date/time of last key modification as date of that last signature.

There is no reliable way of getting the last modification date, because e.g. when reformatting, setting the binding signature date back to the key creation date can be important to ensure older signatures still verify (see the `reformatKey` documentation [here](https://github.com/openpgpjs/openpgpjs/blob/main/src/openpgp.js#L91)). Another special case (pointed out to me by @twiss) might be for people who want short-lived keys, and hence add subkeys in advance that are only valid in the future. Such subkeys might then have a binding signature date that it is also in the future.
For these reasons, we think implementing this type of function needs to be handled by the application directly, so that you can deal with the different cases as you deem appropriate.

> So can you please suggest any other way how to implement this?

I would go through all users and subkeys and verify certification signatures one by one, and look at the creation time if verification is successful (it's similar to what you are doing now, except you manually verify each signature, instead of checking `verified`). As an example on how to do the verification step, you can have a look at `Key.getSigningKey`, which goes through the binding signatures and verifies them: https://github.com/openpgpjs/openpgpjs/blob/main/src/key/key.js#L234 .
You probably want to pass `date = null` to `verify`, otherwise expired keys are always considered invalid. Passing a `null` date also has the effect of considering signature created in the future as valid, which could allow you to detect the "short lived key" case mentioned above, and act accordingly.

**Reply by IvanPizhenko** [reply]
@larabr At the moment I've ended up with following rewrite of the `lastSig()` -  what do you think? (Note: I've read above about the special cases, and this is valid thing, and we will likely have to consider it, but all I want to do at the moment is to have the same functionality with v5 as I've had with v4).

```typescript
  public static lastSig = async (key: Key): Promise<number> => {
    // "await key.getExpirationTime()" no longer works, need some alternate solution
    // discussion is in progress: https://github.com/openpgpjs/openpgpjs/discussions/1491
    const allSignatures: SignaturePacket[] = [];
    for (const user of key.users) {
      const data = { userID: user.userID, userAttribute: user.userAttribute, key: key };
      for (const selfCertification of user.selfCertifications) {
        try {
          await selfCertification.verify(key.keyPacket, enums.signature.certGeneric, data);
          allSignatures.push(selfCertification);
        } catch (e) {
          console.log(e);
        }
      }
    }
    for (const subKey of key.subkeys) {
      try {
        const latestValidSig = await subKey.verify();
        if (latestValidSig) allSignatures.push(latestValidSig);
      } catch (e) {
        console.log(e);
      }
    }
    if (allSignatures.length > 0) {
        allSignatures.sort((a, b) => {
          return (b.created ? b.created.getTime() : 0) - (a.created ? a.created.getTime() : 0);
        });
        const newestSig = allSignatures[0];
        return newestSig.created ? newestSig.created.getTime() : 0;
    }
    throw new Error('No valid signature found in key');
  }
```

**Reply by IvanPizhenko** [reply]
@larabr Btw, found one more inconsistency in the Typescript definitions while writing the above implementation. Please see https://github.com/openpgpjs/openpgpjs/pull/1494

**Comment by IvanPizhenko** [comment]
@larabr Also, if `verified` is considered now "internal", it maybe worth to removed it from typescript definition, because it still there, and this is misleading.
```typescript
export class SignaturePacket extends BasePacket {
...
  public verified: null | boolean;
...
}
```

**Reply by larabr** [reply]
Yeah unfortunately the TS definitions are sort of best-effort and may contain bugs. Feel free to submit a PR to remove the field, adding  the change to #1486 also works.

**Reply by IvanPizhenko** [reply]
Added to #1486
